File: code-library/02-descriptive-stats/descriptive-stats.R
# ==============================================================================
#  MODULE 2: DESCRIPTIVE STATISTICS & TABULATIONS
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Purpose: Summarize continuous and categorical variables, cross-tabulate,
#  compare group means, and examine missing data patterns.
# ==============================================================================

library(tidyverse)
library(scales)      # percent formatting in tables


ces <- readRDS("CES2020_clean.rds")


# ==============================================================================
# SECTION 1: UNIVARIATE SUMMARIES — CONTINUOUS VARIABLES
# ==============================================================================

# summary() gives the five-number summary + mean for every column
ces |>
  select(age, education, ideology5, party_id7, imm_restrict,
         econ_retro, hh_income_change, self_rated_health) |>
  summary()

# More controlled output with summarise + across
# Creates a clean "Table 1" style dataframe you can export
ces |>
  select(age, education, ideology5, party_id7, imm_restrict,
         econ_retro, self_rated_health) |>
  summarise(across(
    everything(),
    list(
      mean   = ~ mean(.x, na.rm = TRUE),
      sd     = ~ sd(.x,   na.rm = TRUE),
      median = ~ median(.x, na.rm = TRUE),
      min    = ~ min(.x,  na.rm = TRUE),
      max    = ~ max(.x,  na.rm = TRUE),
      n      = ~ sum(!is.na(.x))
    ),
    .names = "{.col}__{.fn}"
  )) |>
  pivot_longer(
    everything(),
    names_to  = c("variable", "stat"),
    names_sep = "__"
  ) |>
  pivot_wider(names_from = stat, values_from = value) |>
  mutate(across(where(is.numeric), ~ round(.x, 2)))

# Detailed distribution check — skewness and percentiles
ces |>
  summarise(
    mean   = mean(age, na.rm = TRUE),
    sd     = sd(age, na.rm = TRUE),
    p10    = quantile(age, .10, na.rm = TRUE),
    p25    = quantile(age, .25, na.rm = TRUE),
    median = median(age, na.rm = TRUE),
    p75    = quantile(age, .75, na.rm = TRUE),
    p90    = quantile(age, .90, na.rm = TRUE)
  )


# ==============================================================================
# SECTION 2: FREQUENCY TABLES — CATEGORICAL VARIABLES
# ==============================================================================

# count() produces frequency tables; mutate adds proportions
freq_table <- function(df, var) {
  df |>
    count({{ var }}, .drop = FALSE) |>
    mutate(pct = n / sum(n) * 100) |>
    mutate(pct = round(pct, 1))
}

freq_table(ces, sex)
freq_table(ces, census_region)
freq_table(ces, college)
freq_table(ces, voted)
freq_table(ces, biden_voter)
freq_table(ces, party_id3)
freq_table(ces, party_id7)
freq_table(ces, approve_trump)

# Including NA in counts — always show how much is missing
ces |> count(biden_voter, useNA = "always")
ces |> count(ideology5,   useNA = "always")


# ==============================================================================
# SECTION 3: CROSS-TABULATIONS (TWO-WAY TABLES)
# ==============================================================================

# count(var1, var2) gives raw cell counts
# pivot_wider turns it into a proper crosstab layout

crosstab <- function(df, row_var, col_var) {
  df |>
    count({{ row_var }}, {{ col_var }}) |>
    group_by({{ row_var }}) |>
    mutate(row_pct = round(n / sum(n) * 100, 1)) |>
    ungroup() |>
    pivot_wider(
      names_from  = {{ col_var }},
      values_from = c(n, row_pct),
      names_glue  = "{.value}_{col_var}"
    )
}

# Biden vote by college degree
ces |>
  filter(!is.na(biden_voter), !is.na(college)) |>
  count(college, biden_voter) |>
  group_by(college) |>
  mutate(row_pct = round(n / sum(n) * 100, 1)) |>
  ungroup()

# Biden vote by party ID
ces |>
  filter(!is.na(biden_voter), !is.na(party_id3)) |>
  count(party_id3, biden_voter) |>
  group_by(party_id3) |>
  mutate(row_pct = round(n / sum(n) * 100, 1))

# Biden vote by census region
ces |>
  filter(!is.na(biden_voter), !is.na(census_region)) |>
  count(census_region, biden_voter) |>
  group_by(census_region) |>
  mutate(row_pct = round(n / sum(n) * 100, 1))

# Chi-square test of independence
tbl <- table(ces$college, ces$biden_voter)
chisq.test(tbl)

tbl_region <- table(ces$census_region, ces$voted)
chisq.test(tbl_region)


# ==============================================================================
# SECTION 4: GROUP MEANS
# ==============================================================================

# group_by() + summarise() is the dplyr equivalent of Stata's tabstat by()

# Mean age, ideology, restrictionism by Biden vs Trump voter
ces |>
  filter(!is.na(biden_voter)) |>
  group_by(biden_voter) |>
  summarise(
    mean_age         = round(mean(age, na.rm = TRUE), 2),
    mean_ideology    = round(mean(ideology5, na.rm = TRUE), 2),
    mean_imm_restrict = round(mean(imm_restrict, na.rm = TRUE), 2),
    mean_econ_retro  = round(mean(econ_retro, na.rm = TRUE), 2),
    n                = n()
  )

# Same by party ID (3-category)
ces |>
  filter(!is.na(party_id3)) |>
  group_by(party_id3) |>
  summarise(
    mean_ideology     = round(mean(ideology5, na.rm = TRUE), 2),
    mean_imm_restrict = round(mean(imm_restrict, na.rm = TRUE), 2),
    n                 = n()
  )

# Mean restrictionism by college degree
ces |>
  filter(!is.na(college)) |>
  group_by(college) |>
  summarise(
    mean_restrict = round(mean(imm_restrict, na.rm = TRUE), 2),
    sd_restrict   = round(sd(imm_restrict, na.rm = TRUE), 2),
    n             = sum(!is.na(imm_restrict))
  )


# ==============================================================================
# SECTION 5: T-TESTS — MEAN COMPARISONS
# ==============================================================================

# t.test(y ~ group, data, var.equal = FALSE) — Welch's t-test (default)
# var.equal = FALSE means Welch (unequal variances) — preferred

# Does immigration restrictionism differ by college degree?
t.test(imm_restrict ~ college, data = ces, var.equal = FALSE)

# Does ideology differ between Biden and Trump voters?
t.test(ideology5 ~ biden_voter, data = ces, var.equal = FALSE)

# Does age differ between voters and non-voters?
t.test(age ~ voted, data = ces, var.equal = FALSE)

# Dem vs. Rep only (drop independents for this test)
ces_partisans <- ces |> filter(party_id3 %in% c(1, 2))
t.test(econ_retro ~ party_id3, data = ces_partisans, var.equal = FALSE)


# ==============================================================================
# SECTION 6: CORRELATION MATRIX
# ==============================================================================

# cor() computes Pearson r; use = "pairwise.complete.obs" for pairwise deletion
cor_vars <- ces |>
  select(age, education, ideology5, imm_restrict, econ_retro,
         self_rated_health, biden_voter)

cor_matrix <- cor(cor_vars, use = "pairwise.complete.obs")
round(cor_matrix, 3)

# With p-values using Hmisc (install if needed)
# install.packages("Hmisc")
# library(Hmisc)
# rcorr(as.matrix(cor_vars))

# Or using corrr (tidyverse-adjacent)
# install.packages("corrr")
# library(corrr)
# cor_vars |> correlate() |> shave() |> fashion()


# ==============================================================================
# SECTION 7: MISSING DATA PATTERNS
# ==============================================================================

# Proportion missing for each key variable
ces |>
  select(age, education, ideology5, imm_restrict, voted, biden_voter,
         party_id3, econ_retro) |>
  summarise(across(everything(), ~ mean(is.na(.x)) * 100)) |>
  pivot_longer(everything(),
               names_to = "variable",
               values_to = "pct_missing") |>
  mutate(pct_missing = round(pct_missing, 2)) |>
  arrange(desc(pct_missing))

# Is missingness on ideology5 associated with party?
ces |>
  mutate(miss_ideo = is.na(ideology5)) |>
  group_by(party_id3) |>
  summarise(pct_missing_ideo = mean(miss_ideo, na.rm = TRUE) * 100,
            n = n())


# ==============================================================================
# SECTION 8: WEIGHTED DESCRIPTIVES
# ==============================================================================

# The survey package handles complex survey design properly.
# install.packages("survey")
library(survey)

# Declare survey design (CES post-election weight, no public strata/cluster)
ces_svy <- svydesign(
  ids     = ~1,             # no clustering variable publicly available
  weights = ~wt_post,
  data    = ces
)

# Weighted means
svymean(~ age + college + voted + ideology5, design = ces_svy, na.rm = TRUE)

# Unweighted for comparison
ces |>
  summarise(
    mean_age     = mean(age, na.rm = TRUE),
    mean_college = mean(college, na.rm = TRUE),
    mean_voted   = mean(voted, na.rm = TRUE),
    mean_ideo    = mean(ideology5, na.rm = TRUE)
  )

message("Module 2 complete.")

File: code-library/02-descriptive-stats/descriptive-stats.py
# ==============================================================================
#  MODULE 2: DESCRIPTIVE STATISTICS & TABULATIONS
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Stack: pandas, scipy (t-tests), numpy
# ==============================================================================

import pandas as pd
import numpy as np
from scipy import stats

ces = pd.read_parquet("CES2020_clean.parquet")


# ==============================================================================
# SECTION 1: UNIVARIATE SUMMARIES — CONTINUOUS VARIABLES
# ==============================================================================

# DataFrame.describe() gives count, mean, std, min, quartiles, max
cont_vars = ["age", "education", "ideology5", "party_id7",
             "imm_restrict", "econ_retro", "self_rated_health"]

print(ces[cont_vars].describe().round(2))

# More detailed: percentiles
print(ces["age"].describe(percentiles=[.1, .25, .5, .75, .9]).round(2))

# Custom Table 1 — mean, SD, median, N for each variable
def table1_row(series):
    return pd.Series({
        "Mean":   series.mean(),
        "SD":     series.std(),
        "Median": series.median(),
        "Min":    series.min(),
        "Max":    series.max(),
        "N":      series.notna().sum(),
    })

table1 = ces[cont_vars].apply(table1_row).T.round(2)
print(table1)


# ==============================================================================
# SECTION 2: FREQUENCY TABLES — CATEGORICAL VARIABLES
# ==============================================================================

# value_counts() gives counts; normalize=True gives proportions
# dropna=False includes NaN in the count

def freq_table(series):
    counts = series.value_counts(dropna=False).sort_index()
    pct    = series.value_counts(normalize=True, dropna=False).sort_index() * 100
    return pd.DataFrame({"n": counts, "pct": pct.round(1)})

for col in ["sex", "census_region", "college", "voted",
            "biden_voter", "party_id3", "party_id7"]:
    print(f"\n--- {col} ---")
    print(freq_table(ces[col]))


# ==============================================================================
# SECTION 3: CROSS-TABULATIONS
# ==============================================================================

# pd.crosstab() produces a two-way frequency table
# normalize="index" gives row percentages

# Biden vote by college degree
ct = pd.crosstab(ces["college"], ces["biden_voter"],
                 margins=True, dropna=False)
print("\nBiden vote × College:\n", ct)

# Row percentages
ct_pct = pd.crosstab(ces["college"], ces["biden_voter"],
                     normalize="index") * 100
print(ct_pct.round(1))

# Chi-square test of independence
tbl = pd.crosstab(ces["college"].dropna(),
                  ces["biden_voter"].dropna())
chi2, p, dof, expected = stats.chi2_contingency(tbl)
print(f"\nChi2={chi2:.2f}, df={dof}, p={p:.4f}")

# Biden vote by party ID
print("\nBiden vote × Party ID (row %):")
print(
    pd.crosstab(ces["party_id3"], ces["biden_voter"],
                normalize="index").round(3) * 100
)

# Turnout by region
print("\nVoted × Region (row %):")
print(
    pd.crosstab(ces["census_region"], ces["voted"],
                normalize="index").round(3) * 100
)


# ==============================================================================
# SECTION 4: GROUP MEANS
# ==============================================================================

# groupby().agg() computes multiple statistics per group

group_means = (
    ces.groupby("biden_voter")[["age", "ideology5", "imm_restrict", "econ_retro"]]
    .agg(["mean", "std", "count"])
    .round(2)
)
print("\nGroup means by Biden vote:\n", group_means)

# By party ID
print("\nGroup means by party:")
print(
    ces.groupby("party_id3")[["ideology5", "imm_restrict"]]
    .mean()
    .round(2)
)

# By college
print("\nRestriction by college:")
print(
    ces.groupby("college")["imm_restrict"]
    .agg(["mean", "std", "count"])
    .round(2)
)


# ==============================================================================
# SECTION 5: T-TESTS — MEAN COMPARISONS
# ==============================================================================

# scipy.stats.ttest_ind() runs an independent-samples t-test
# equal_var=False → Welch's t-test (default in most social science)

def welch_ttest(df, outcome, group_col, group_vals):
    g1 = df.loc[df[group_col] == group_vals[0], outcome].dropna()
    g2 = df.loc[df[group_col] == group_vals[1], outcome].dropna()
    t, p = stats.ttest_ind(g1, g2, equal_var=False)
    print(f"{outcome} by {group_col} ({group_vals[0]} vs {group_vals[1]}): "
          f"t={t:.3f}, p={p:.4f}, "
          f"means=[{g1.mean():.2f}, {g2.mean():.2f}]")

welch_ttest(ces, "imm_restrict", "college", [0, 1])
welch_ttest(ces, "ideology5",    "biden_voter", [1, 0])
welch_ttest(ces, "age",          "voted",    [1, 0])

# Dem vs. Rep only (drop independents)
ces_partisans = ces[ces["party_id3"].isin([1, 2])]
welch_ttest(ces_partisans, "econ_retro", "party_id3", [1, 2])


# ==============================================================================
# SECTION 6: CORRELATION MATRIX
# ==============================================================================

# DataFrame.corr() computes Pearson correlations (pairwise complete obs)

corr_vars = ["age", "education", "ideology5", "imm_restrict",
             "econ_retro", "self_rated_health", "biden_voter"]

corr_matrix = ces[corr_vars].corr(method="pearson")
print("\nCorrelation matrix:\n", corr_matrix.round(3))

# With p-values (manual computation)
def corr_pvalue(df, var1, var2):
    valid = df[[var1, var2]].dropna()
    r, p  = stats.pearsonr(valid[var1], valid[var2])
    return round(r, 3), round(p, 4)

for v in ["education", "ideology5", "imm_restrict", "econ_retro"]:
    r, p = corr_pvalue(ces, "biden_voter", v)
    print(f"biden_voter × {v}: r={r}, p={p}")


# ==============================================================================
# SECTION 7: MISSING DATA PATTERNS
# ==============================================================================

key_vars = ["age", "education", "ideology5", "party_id3",
            "imm_restrict", "voted", "biden_voter"]

missing = (
    ces[key_vars]
    .isna()
    .mean()
    .mul(100)
    .round(2)
    .rename("pct_missing")
    .reset_index()
    .rename(columns={"index": "variable"})
    .sort_values("pct_missing", ascending=False)
)
print("\nMissing data (%):\n", missing)

# Is missingness on ideology5 associated with party?
miss_by_party = (
    ces.assign(miss_ideo=ces["ideology5"].isna())
    .groupby("party_id3")["miss_ideo"]
    .mean()
    .mul(100)
    .round(2)
)
print("\n% missing ideology by party:\n", miss_by_party)


# ==============================================================================
# SECTION 8: WEIGHTED DESCRIPTIVES
# ==============================================================================

# pandas doesn't have native survey-weight support for all statistics.
# Use numpy.average(values, weights=weights) for weighted means.

def weighted_mean(series, weights):
    mask = series.notna() & weights.notna()
    return np.average(series[mask], weights=weights[mask])

print("\nWeighted vs unweighted means:")
for col in ["age", "college", "voted", "ideology5"]:
    unwt = ces[col].mean()
    wt   = weighted_mean(ces[col], ces["wt_post"])
    print(f"  {col}: unweighted={unwt:.3f}, weighted={wt:.3f}")

# For full survey-weighted inference (SEs, regression), use the statsmodels
# WLS or the linearmodels package — demonstrated in Module 5.

print("\nModule 2 complete.")
